### PR TITLE
Add ISAT functionality to dev

### PR DESCRIPTION
**Pull request by blttkgl**
Hello,

At Wärtsilä we have been working on extending the DLBFoam to work together with ISAT, as well as extending ISAT to work with the jacobian generated with pyJac. 

The main principle is that before any balancing operation, the tabulation process is executed and problems are sorted into ones that can be retrieved from the binary tree, and the ones that can't. The balancing then happens only on the non-retrieved problems. Depending on the application, this provides 20 to 50% speedup (Please check the thesis work by Daniel Virokannas: https://aaltodoc.aalto.fi/server/api/core/bitstreams/d99091dd-25cb-4e66-9233-06cbee0229ac/content)

I am creating this pull request if this is of interest to you guys, as well as in an effort to sync up our shared progress on DLBFoam. I'd be happy to provide a more detailed breakdown of the implementation.

Bulut

### CODE REVIEW

**Comment by moreff** [comment]
Hi, thank you for your effort! It seems to be working according to my tests.

However, there are some minor issues:
1. Multiple OpenFOAM classes are directly copy-pasted and renamed. This complicates the maintenance of the code unnecessarily. Inheriting ISAT_pyJac from ISAT directly would remove the need to copy-paste other ISAT classes.
2. The reduction switch in chemistry can be checked upon the construction of ISAT_pyJac and an error message can be printed saying that reduction is not implemented.
3. Tutorial for DLB+pyJac was overwritten. I don't see a reason to do that, there can be both tutorials with and without ISAT, since ISAT does not have to be used with pyJac. Just like in OpenFOAM they have multiple tutorials for different solver configurations, we should have the same I think.
4. What did you decide on this discussion: https://github.com/Aalto-CFD/DLBFoam/issues/38#issuecomment-2138789964 ?
5. Also, it would be easier to review if there weren't so many files/lines "modified" with spaces or new lines added, or swapped... It also pollutes the repo history, showing that some files were modified with a commit when in reality they weren't. I believe, at least line change and space addition ignoring can be configured in a git client.

Unfortunately, your commits cannot be rebased to the repo due to conflicts. I can squash them into a single commit and merge them, and they will be shown as a single commit. 

I can take care of 1 and 2 after I accept the pull request. 
Regarding 5, no changes are needed, just in future to ease the review for others it would be good to avoid these. 

What do you think?

File: src/thermophysicalModels/chemistryModel/Make/options
Comment (moreff) [inline src/thermophysicalModels/chemistryModel/Make/options:1]: This change is not needed

**Comment by blttkgl** [comment]
Hey,

Thank you for the comments, you raise a lot of valid points. 

1- I agree, if I recall correctly I opted to this quick and dirty solution because I needed the functionality, and it was cumbersome the figure out the inheritance since there needs to be quite a lot of changes due to the different shape and size of the thermophysical vector. It would be great if you have the time to make this work, its definitely the better approach.

2- Agreed. I very much never use it, but it would be also rather straightforward to add reduction to the dlbfoam (not dlbfoam_pyjac) if you guys need it for any reason.

3- You could perhaps just ignore those upcoming changes? As long as the tutorials are consistent with the latest OpenFOAM and uses foamRun it shouldn't need any changes.

4- Hasan's suggestion is the correct one. Would you like me to make that change in my branch while this request is open so it is added to the pull request?

5- Agreed! We implemented these changes swiftly to get to work, and unfortunately I did not pay much attention to the touches to the other files. Do you see a way to get rid of them? 

Bulut

**Comment by moreff** [comment]
1. Ok, I will look into it.
2. I meant not the implementation of the reduction, but the removal of all the code related to the reduction switch that you did. The removal is not needed if there is a check in the constructor, thus we can re-use the OpenFOAM code as it is and again simplify maintenance. I will do that. Regarding reduction implementation, there are no such plans currently.
3. I cannot ignore them, there is no such functionality in GitHub to my knowledge. I have to accept **all** the changes that you push. So it is either a creator of a pull request that changes the contents of the pull request, or the repo owner has to revert things basically manually going through all the files **after** the pull request is accepted, creating new commits. Anyway, I will restore the tutorials manually, it will be faster.
4. I can do it along with other changes.
5. At this point, I don't know any easy solution (there might be though). Another thing why this is important is that it can potentially cause problems with commands like `git cherry-pick`, which I use to maintain older versions of the repository, stored in other branches. So let's just try to avoid this in future to avoid extra work from my side.

**Comment by moreff** [comment]
Hi @blttkgl , upon closer look I noticed that there is a large performance difference between the original ISAT implementation in OpenFOAM and yours. With loadbalanced chemistry (without pyJac) in theory on a single-cell setup they should have the same performance and results, or at least close, right? 
But they are not, according to my tests. In the attached minimum example your implementation takes 6 times more steps.
[testISAT.txt](https://github.com/user-attachments/files/16156485/testISAT.txt)
Since you are more familiar with the implementation, do you have an explanation?

**Comment by blttkgl** [comment]
Hey @moreff ,

Yes, I managed to recreate this on my end. It's odd, I'll look into it and let you know. You're right, the behavior should be identical.

At the same time, if you run a 2D tutorial with ISAT, using standard or DLBFoam, the performance is the same:

[testISAT_CounterFlow.txt](https://github.com/user-attachments/files/16157016/testISAT_CounterFlow.txt)

Could you test this and confirm, while I look into this?

**Comment by blttkgl** [comment]
Also I wonder if chemFoam is the right solver to test this, even though I agree with you that the results should be identical (and they are if the tabulation entry is removed).

Because chemFoam simply called the solve() method from the chemistry object, and doesn't really do any sort of tabulation, since there is only one cell and nothing to be tabulated, right? 

https://github.com/OpenFOAM/OpenFOAM-dev/blob/996d26811387b4974989c29a8ece1471bacf15f4/applications/solvers/chemFoam/solveChemistry.H

**Comment by moreff** [comment]
Yes, with counter-flow flame the performance is the same.  
But the tabulation is done is `chemistry.solve()`. It does tabulate the solution initially and at least at initial steps it uses it I think.
In e.g. `add_isat.out` log I see that the original implementation tabulates once in the beginning of the simulaiton, whereas DLB does it five times.  
Of course, this is a limiting weird case where tabulation should never be used. By I tried it as a simplest check if the implementation is consistent with OpenFOAM's.

**Comment by blttkgl** [comment]
The issue is that the internal time steppings are different between DLB and standard when tabulation is used, if you check the logs. I'll try to find the root cause.

**Comment by blttkgl** [comment]
Hey @moreff 

In short, the difference between two models is that the deltaTMin returned by the ::solve in standard and DLBFoam is different. I have yet to find the root cause, but the performance and results of these two approaches on other tutorials seem to work fine. 

Regarding the chemFoam, I do not think it makes sense to use tabulation there. There is only one cell, and if that is tabulated/retrieve it would create a closed loop where the same problem is being tabulated and retrieved onto itself. Without involving DLBFoam, if you run the gri example with tabulation and plot the temperature profile, you see that the ignition is not captured when tabulation is active:
GRI example
image

Same example with tabulation
OF_vs_CHEMKINII

If we could recreate the issue on an actual foamRun case, such as counterFlowFlame2D_GRI tutorial, I think that would make it easier to debug it.

Let me know what you think,

Bulut

**Comment by moreff** [comment]
Yes, I said above that tabulation in principle should not be used for this kind of problem (autoignition). Also, OpenFOAM's default configuration of the ISAT is very coarse, it is not surprising that validation fails. When I set the tabulation tolerance to 1e-5 I get a good match for OpenFOAM, but a mismatch for DLB.

I view this 0d case as a consistency test, which indicates that there are some differences, which is unexpected in my understanding. I think it is good to use this 0d case to catch problems, which might be hard to catch in larger scales. Even though the cell tabulates its own chemistry solution, I don't see a problem here, it still should work the same between OpenFOAM and DLB models. 

I wasn't able to recreate it in a larger case so far.

I have doubts about these parts of the code:
https://github.com/OpenFOAM/OpenFOAM-dev/blob/master/src/thermophysicalModels/chemistryModel/chemistryModel/chemistryModel.C#L651
https://github.com/Aalto-CFD/DLBFoam/blob/v1.1_OFdev/src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C#L499

It looks like OpenFOAM takes the temperature and pressure from the table and in DLB it is not happening. Do I understand it right?

**Comment by blttkgl** [comment]
Hey Ilya,
https://github.com/Aalto-CFD/DLBFoam/blob/0ac8b334aeff4d885c7fb01a4ea46dd5698c92e7/src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C#L226

The ISAT-DLB logic works as follows:

`    if(problem.procNo == Pstream::myProcNo())`
- Binary trees are checked only if the problem is in its original rank, we do not want guest problems to interact with the binary trees (due to cell indexing issues, which I hope to fix in the future).

https://github.com/Aalto-CFD/DLBFoam/blob/0ac8b334aeff4d885c7fb01a4ea46dd5698c92e7/src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C#L231

- If we can retrieve the problem from the binary tree, we do not do anything else, just update the problem.c value here
https://github.com/Aalto-CFD/DLBFoam/blob/0ac8b334aeff4d885c7fb01a4ea46dd5698c92e7/src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C#L505

This problem.c is then used to update the reaction rate:

https://github.com/Aalto-CFD/DLBFoam/blob/0ac8b334aeff4d885c7fb01a4ea46dd5698c92e7/src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C#L268

I think I left those assignments out, because temperature and pressure is not updated on the chemistry model level. Chemistry model calculates the reaction rate, which then is added as a source term to the energy equation, which calculates the temperature via ideal gas law. 

I think you pointed out a very interesting and simple test which we may have overlooked. For us we have performed benchmarks on our baseline case, which gave the same results with standard, and we moved further. Let's see if we can pinpoint the issue.

**Comment by moreff** [comment]
Thank you for the explanation! You are right, they can be left out.  

Another thing is that in DLB chemistry time step is always updated
https://github.com/Aalto-CFD/DLBFoam/blob/0ac8b334aeff4d885c7fb01a4ea46dd5698c92e7/src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C#L481
whereas in the original implementation it is updated only when a new solution is obtained:  
https://github.com/OpenFOAM/OpenFOAM-dev/blob/61ae26e5f739b2d4242d8b4e7b6846c0e0e1c07b/src/thermophysicalModels/chemistryModel/chemistryModel/chemistryModel.C#L757

To my understanding, tabulation does not distinguish between the type of variables it is used with, so it will tabulate and scale also this time step.

**Pull request merged**